### PR TITLE
Fix some issues with periodic boundary condition with cfd-dem

### PR DESCRIPTION
**Pull request by acdaigneault**
# Description of the problem

This pull request fixes 3 issues related to periodic boundary conditions feature with CFD-DEM coupling with the VANS solver.

1. Periodic boundary condition feature wasn't implemented in CFD-DEM
2. Having a particle crossing a periodic wall with 2 cells not on the same processor was not working
3. An issue has found using DEM with custom mesh having diamond cells

# Description of the solution

The solutions' number are linked to the previous issues as listed as above.

1. PBC was added to cfd-dem in the cfd-dem coupling part as in the dem solver (see #557)
2. A condition related to ghost cells was added prior to get data of the periodic cell while mapping pair of cells on the periodic boundary (see periodic_boundaries_manipulator.cc, line 75)
3. Usage of GMSH mesh with diamond thrown exception (`map::at`) which was related to using outlet boundaries (or periodic) with diamond cells. This error was because adding line boundaries to boundaries information was looking for normal vector of face neighbors, but face neighbors at outlet or periodic are not stored in boundary_cells_information map at the search step. The solution was to add a check step if one of the faces on boundary was on a outlet or a periodic boundary (see find_boundary_cells_information.cc, line 573) Some refactoring was done in this function.

Note : Another issue was found while using `expand particle-wall contact search` with outlet or periodic boundary. It also thrown a `map::at` error. It is not fixed in this PR, but will in another one.

# How Has This Been Tested?

- Application test with GMSH mesh with diamond cell on boundaries was added

# Future changes

- Fix the `expand particle-wall contact search` feature with outlet or periodic boundary condition issue

### CODE REVIEW

File: applications_tests/dem_3d/periodic_boundary_gmsh.prm [hunk 1]
```diff
@@ -0,0 +1,100 @@
+# Listing of Parameters
+# ---------------------
+
+#---------------------------------------------------
+# Simulation and IO Control
+#---------------------------------------------------
+subsection simulation control
+    set time step                = 1.5e-6
```
Comment (tonielgeitani) [inline applications_tests/dem_3d/periodic_boundary_gmsh.prm:8]: Why is the timestep the same as the end time? So you are only doing one iteration?

File: applications_tests/dem_3d/periodic_boundary_gmsh.prm [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+end
+
+# --------------------------------------------------
+# Model parameters
+#---------------------------------------------------
+subsection model parameters
+  set contact detection method 		   		 = dynamic
+  set load balance method				     = dynamic
+  set load balance threshold				 = 0.5
+  set dynamic load balance check frequency	 = 10000
```
Comment (tonielgeitani) [inline applications_tests/dem_3d/periodic_boundary_gmsh.prm:28]: Based on your end time you are not doing any load balancing, so its better to remove the section if not using it.

File: applications_tests/dem_3d/periodic_boundary_gmsh.output [hunk 3]
```diff
@@ -0,0 +1,24 @@
+Running on 1 rank(s)
+*****************************************************************
+
+DEM time-step is 7.71088% of Rayleigh time step
+Reading triangulation
+
+Finished reading triangulation
+
+Warning: There are diamond-shaped cells in the input triangulation. It is strongly recommended to use a different triangulation without such cells. It should be mentioned that these cells are not detected if you have grid motion
+Warning: expansion of particle-wall contact list is disabled.
```
Comment (tonielgeitani) [inline applications_tests/dem_3d/periodic_boundary_gmsh.output:10]: expansion of particle should be on since your geometry gmsh is a pipe

**Review by tonielgeitani — APPROVED** [review]

File: applications_tests/dem_3d/periodic_boundary_gmsh.output (hunk 3 above)
Comment (acdaigneault) [reply applications_tests/dem_3d/periodic_boundary_gmsh.output:10]: This is the feature that is currently not working with periodic or outlet boundaries with DEM unfortunately. Will work on that in next PR.

**Comment by acdaigneault** [comment]
@blaisb, do you think you could merge the PR without deleting the branch so I could fix the `expand particle-wall contact search` thing in the same branch? Thanks

File: applications_tests/dem_3d/insert_periodic_boundary_gmsh.prm [hunk 4]
```diff
@@ -0,0 +1,91 @@
+# Listing of Parameters
+# ---------------------
+
+#---------------------------------------------------
```
Comment (blaisb) [inline applications_tests/dem_3d/insert_periodic_boundary_gmsh.prm:4]: Have you indented the file with Paul's script?

File: source/dem/find_boundary_cells_information.cc [hunk 5]
```diff
@@ -547,37 +546,59 @@ BoundaryCellsInformation<dim>::add_cells_with_boundary_lines_to_boundary_cells(
...
+                                       outlet_boundaries.end() &&
+                                     !cell_one->has_periodic_neighbor(
+                                       face_id_one));
+                                  face_two_is_wall =
+                                    (std::find(outlet_boundaries.begin(),
+                                               outlet_boundaries.end(),
+                                               face_two->boundary_id()) ==
+                                       outlet_boundaries.end() &&
+                                     !cell_two->has_periodic_neighbor(
+                                       face_id_two));
```
Comment (blaisb) [inline source/dem/find_boundary_cells_information.cc:587]: Good way to check this.

**Review by blaisb — APPROVED** [review]
Very good. very few comments. Can you just squash your ocmmits and apply our comments and this will be ready for a merger

File: applications_tests/dem_3d/insert_periodic_boundary_gmsh.prm (hunk 4 above)
Comment (acdaigneault) [reply applications_tests/dem_3d/insert_periodic_boundary_gmsh.prm:4]: No, does it have documentation to use it?

**Comment by acdaigneault** [comment]
> > @blaisb, do you think you could merge the PR without deleting the branch so I could fix the `expand particle-wall contact search` thing in the same branch? Thanks
> 
> Yes, but you will need to squash your commits before i merge so that I merge a single commit instead of multiple commits.

I did merge master into the branch between some commits and I don't know how to deal with squashing commits in that situation. So you can delete the branch, I will start a new one. Thanks :)
